packages/auth/src/mfa/assertions/totp.ts

```diff
@@ -67,12 +84,162 @@ export class TotpMultiFactorGenerator {
    * @returns A promise to {@link TotpSecret}.
    */
   static async generateSecret(
-    _session: MultiFactorSession
+    session: MultiFactorSession
   ): Promise<TotpSecret> {
-    throw new Error('Unimplemented');
+    const mfaSession = session as MultiFactorSessionImpl;
+    _assert(
+      typeof mfaSession.auth !== 'undefined',
+      AuthErrorCode.INTERNAL_ERROR
+    );
+    const response = await startEnrollTotpMfa(mfaSession.auth, {
+      idToken: mfaSession.credential,
+      totpEnrollmentInfo: {}
+    });
+    return TotpSecret.fromStartTotpMfaEnrollmentResponse(
+      response,
+      mfaSession.auth
+    );
   }
+
   /**
    * The identifier of the TOTP second factor: `totp`.
    */
   static FACTOR_ID = FactorId.TOTP;
 }
+
+export class TotpMultiFactorAssertionImpl
+  extends MultiFactorAssertionImpl
+  implements TotpMultiFactorAssertion
+{
+  constructor(
+    readonly otp: string,
+    readonly enrollmentId?: string,
+    readonly secret?: TotpSecret
+  ) {
+    super(FactorId.TOTP);
+  }
+
+  static _fromSecret(
+    secret: TotpSecret,
+    otp: string
+  ): TotpMultiFactorAssertionImpl {
+    return new TotpMultiFactorAssertionImpl(
+      (otp = otp),
```

Comment:
what does this assignment do?

Reply:
I added this to selectively pass in arguments, but that isn't supported. Removed this.